jax_galsim/moffat.py

```diff
@@ -0,0 +1,467 @@
+from functools import partial
+
+import jax
+import jax.numpy as jnp
+
+import jax.scipy as jsc
+
+
+from jax_galsim.gsobject import GSObject
+from jax_galsim.gsparams import GSParams
+from jax_galsim.core.draw import draw_by_xValue, draw_by_kValue
+
+from jax_galsim.core.bessel import j0
+from jax_galsim.core.integrate import ClenshawCurtisQuad, quad_integral
+
+import galsim as _galsim
+from jax._src.numpy.util import _wraps
+from jax.tree_util import register_pytree_node_class
+
+import tensorflow_probability as tfp
+
+
+@jax.jit
+def _Knu(nu, x):
+    """Modified Bessel 2nd kind for Untruncated Moffat"""
+    return tfp.substrates.jax.math.bessel_kve(nu * 1.0, x) / jnp.exp(jnp.abs(x))
+
+
+@jax.jit
+def MoffatIntegrant(x, k, beta):
+    """For truncated Hankel used in truncated Moffat"""
+    return x * jnp.power(1 + x**2, -beta) * j0(k * x)
+
+
+def _xMoffatIntegrant(k, beta, rmax, quad):
+    return quad_integral(partial(MoffatIntegrant, k=k, beta=beta), 0.0, rmax, quad)
+
+
+@jax.jit
+def _hankel(k, beta, rmax):
+    return jax.vmap(
+        partial(
+            _xMoffatIntegrant, beta=beta, rmax=rmax, quad=ClenshawCurtisQuad.init(150)
+        )
+    )(k)
+
+
+def MoffatCalculateSRFromHLR(re, rm, beta, Nloop=1000):
+    """
+    The basic equation that is relevant here is the flux of a Moffat profile
+    out to some radius.
+
+    flux(R) = int( (1+r^2/rd^2 )^(-beta) 2pi r dr, r=0..R )
+            = (pi rd^2 / (beta-1)) (1 - (1+R^2/rd^2)^(1-beta) )
+    For now, we can ignore the first factor.  We call the second factor fluxfactor below,
+    or in this function f(R).
+    We are given two values of R for which we know that the ratio of their fluxes is 1/2:
+    f(re) = 0.5 * f(rm)
+
+    nb1. rd aka r0 aka the scale radius
+    nb2. In GalSim definition rm = 0 (ex. no truncated Moffat) means in reality rm=+Inf.
+         BUT the case rm==0 is already done, so HERE rm != 0
+    """
+    # JEC Where to do these checking?
+    # assert rm != 0.0, f"MoffatCalculateSRFromHLR: rm=={rm} should be done elsewhere"
+    #
+    # assert (
+    #    rm > jnp.sqrt(2.0) * re
+    # ), f"MoffatCalculateSRFromHLR: Cannot find a scaled radius: rm={rm}, sqrt(2)*re={jnp.sqrt(2.) * re}"
```

Comment:
I wouldn't leave this as a comment... Could you make an issue stating that we should add these checks back later? Not urgent if it was required to passed the galsim tests

Reply:
once you do that, pleaes delete this commented out code

Reply:
done